### PR TITLE
`typeof` on a well-known symbol in a computed property name has a poor error

### LINKED ISSUE #42523: `typeof` on a well-known symbol in a computed property name has a poor error
Found talking to @benlesh

If you try to write the following

```ts
interface SymbolConstructor {
  observable: symbol;
}

declare function from<T>(obj: { [typeof Symbol.observable](): Observable<T> }): Observable<T>;
//                               ^^^^^^ Look at this use of 'typeof'.

type Observer<T> = (x: T) => void;

interface Observable<T> {
  subscribe(observer: Observer<T>): { unsubscribe(): void }
}
```

You'll get an error like

```
A computed property name in a type literal must refer to an expression whose type is a literal type or a 'unique symbol' type.
```

This is pretty confusing, because we *do* support well-known symbols that are declared with `symbol` instead of unique symbol if you use the following syntax:

```
declare function from<T>(obj: { [Symbol.observable](): Observable<T> }): Observable<T>;
//                               ^ No 'typeof'.
```

It would be great if we could detect this case and specialize the error message with something like

```
Computed property names can only be named with well-known symbols without the 'typeof' operator. Try replacing `typeof {0}` with `{0}`.
```

### PR DESCRIPTION
**Pull request by jonhue**
Fixes #42523

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -40414,6 +40414,9 @@ namespace ts {
                 return checkGrammarForInvalidDynamicName(node.name, Diagnostics.A_computed_property_name_in_an_interface_must_refer_to_an_expression_whose_type_is_a_literal_type_or_a_unique_symbol_type);
             }
             else if (node.parent.kind === SyntaxKind.TypeLiteral) {
+                if (node.name.kind === SyntaxKind.ComputedPropertyName && node.name.expression.kind === SyntaxKind.TypeOfExpression) {
+                    return checkGrammarForInvalidDynamicName(node.name, Diagnostics.Computed_property_names_can_only_be_named_with_well_known_symbols_without_the_typeof_operator_Try_replacing_typeof_0_with_0, (node.name.expression as TypeOfExpression).expression);
```
Comment (jonhue) [inline src/compiler/checker.ts:40418]: Do we have a function to print the text associated with a node in the syntax tree? Right now, `node.name.expression.expression` refers to a node that would want to print the text of.

Also, do you see why the node is not automagically narrowed to `TypeOfExpression`? 😄

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -6342,5 +6342,9 @@
     "Invalid value for 'jsxFragmentFactory'. '{0}' is not a valid identifier or qualified-name.": {
         "category": "Error",
         "code": 18035
+    },
+    "Computed property names can only be named with well-known symbols without the 'typeof' operator. Try replacing 'typeof {0}' with '{0}'.": {
+        "category": "Error",
+        "code": 18036
     }
```
Comment (jonhue) [inline src/compiler/diagnosticMessages.json:6349]: I wasn't super sure where to put this. Do we always put new errors at the end of this file with an incremented status code or should I move this closer to the related errors?

File: tests/baselines/reference/computedPropertyNamesWithTypeOfOperator.errors.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    
+    interface Observable<T> {
+      subscribe(observer: Observer<T>): { unsubscribe(): void }
+    }
+    
+    declare function from1<T>(obj: { [typeof Symbol.observable](): Observable<T> }): Observable<T>;
+                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS18036: Computed property names can only be named with well-known symbols without the 'typeof' operator. Try replacing 'typeof [object Object]' with '[object Object]'.
+                                             ~~~~~~
+!!! error TS2585: 'Symbol' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
```
Comment (jonhue) [inline tests/baselines/reference/computedPropertyNamesWithTypeOfOperator.errors.txt:21]: I'm also not sure where this error is coming from. When I run type inference on this file outside of the tests, it doesn't pop up 🤔

File: src/compiler/diagnosticMessages.json (hunk 2 above)
Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:6349]: We do try to put them in the closest section to a relevant error typically

File: tests/baselines/reference/computedPropertyNamesWithTypeOfOperator.errors.txt (hunk 3 above)
Comment (DanielRosenwasser) [reply tests/baselines/reference/computedPropertyNamesWithTypeOfOperator.errors.txt:21]: You'll need to include es2015.symbol as part of the `lib`. `SymbolConstructor` isn't used in the base lib.

In tests, you can do this with `// @lib: es5, es2015.symbol` if I recall correctly.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Thanks for the PR @jonhue! However, since I filed that issue yesterday, I've realized that maybe https://github.com/microsoft/TypeScript/pull/24738 is a better direction, where the type-checker will automatically promote well-known symbols to being unique. If we merge this in, we may have to back-out this change afterwards. 😬

**Comment by jonhue** [comment]
No worries @DanielRosenwasser! That stuff happens 😄 

There is still one open question (https://github.com/microsoft/TypeScript/pull/42530#discussion_r565906651) concerning how to get the text represented by a node.
Happy to close this or follow-up on this depending on what you think is the best idea here!

**Comment by DanielRosenwasser** [comment]
I think we have something like `tryGetPropertyAccessOrIdentifierToString` you can use in the future.

**Comment by jonhue** [comment]
@DanielRosenwasser are you able to see why the Travis tests are failing? It looks like all other tests are successful 🤔

**Comment by DanielRosenwasser** [comment]
Closing this based on the other PR. We ripped Travis out because of failing tests btw.

**Pull request closed without merging**